### DISCUSSION TITLE
Ask before saving to and loading from a global state

### DISCUSSION THREAD

**Discussion by ghost** [grumpycoders/pcsx-redux#1119] [Ideas]
Because both "Save global state" and "Load global state" are **NEXT TO** each other in the File menu **AND** they are both **SMALL** in the screen, there is a **BIG** chance and probability that the user will both accidentally and mistakenly "Save global state" instead of "Load global state" and vice versa when his or her hand **SLIPS**!

So my idea, proposal and suggestion is to ask the user the following question:

> Are you sure that you want to save to global state?

Every time the user wants to "Save global state"

AND

> Are you sure that you want to load from global state?

Every time the user wants to "Load global state"

**NOTE:**

No dialog appears if the user saves and loads states to and from **slots**, Not even if he or she **quickly** saves or loads a state by pressing a **key** on his or her **keyboard** because the **slots** menus are designed in a way so the chances and probabilities for an accident and a mistake are too small and the keyboard is used to **quickly** save and load a state and a dialog will make it both **not** quick and **annoying**.

These two dialogs are **only** for "Save global state" and "Load global state" **specifically**!

Is that clear? If yes then what do you think about this idea? I am will be glad to hear your opinions!

Optionally you can add a new checkbox so the user can choose to **not** show these dialogs even for "Save global state" and "Load global state"

**Comment by wheremyfoodat** [comment]
But like, you need to spawn the top menu bar, press the file button, and then click one of those options to do it. I can't see many people doing it accidentally.

**Reply by ghost** [reply]
> But like, you need to spawn the top menu bar, press the file button, and then click one of those options to do it. I can't see many people doing it accidentally.

So you are not going to make this dialog at all. Never. Okay. I will compile the source,  add it myself and make my own custom build and not share it with anyone then.

And I don't think that you understand what I am talking about at all.

Spawning the top menu bar and pressing the file button are done on purpose but clicking  on the "Load global state" button can be done on accident if the hand of the user slipped down slightly and moved the mouse cursor down enough to point at the "Load global state" button instead of the "Save global state" button that the user intended to click.

This accident is not so common but still may happen sometimes and this already happened to me several times by the way and this is the reason that I started this discussion at all.

Even worse if the user made a mistake in the game, like making a wrong choice or not pressing a button on time, and he or she wants to undo his or her mistake by going to "Load global state" but his or her hand slipped and accidentally clicked on the "Save global state" instead and saved the mistake that he or she did in the game and he or she can no longer undo it.

This is what I was trying to say. Your conclusion is wrong.

Of course that doing all the spawning the top menu bar, pressing the file button and clicking "Load global state" instead of "Save global state" all by accident is impossible.

Perhaps I had to explain myself better.

Is that clear enough?

Anyway I will take a vacation. I need a rest.

**Reply by johnbaumann** [reply]
PR's welcome

**Reply by ghost** [reply]
> PR's welcome

What does PR mean exactly?

**Reply by NDR008** [reply]
A PR= pull request. Actually making your own branch/fork, making the changes and sharing back (by means of a PR) is the best way to try and add a feature that you thought was worth adding and already did for themselves, they can share up to the original source.

**Reply by ghost** [reply]
> A PR= pull request. Actually making your own branch/fork, making the changes and sharing back (by means of a PR) is the best way to try and add a feature that you thought was worth adding and already did for themselves, they can share up to the original source.

Now I get it. I better do it before more users like me will accidentally lose progress in their game or save mistakes that they did in the game.

**Reply by ghost** [reply]
> But like, you need to spawn the top menu bar, press the file button, and then click one of those options to do it. I can't see many people doing it accidentally.

Interesting what you are saying to me but please tell me why [Chidreams Emulation GamePlay](https://www.youtube.com/@ChidreamsEmulationGamePlay) in his new [video](https://www.youtube.com/watch?v=cRPPS4BzAW4) between 3:10 to 3:17 **accidentally** selected the wrong process Cheat Engine itself instead of rpcs3? And he admits that he did this mistake at 5:25!

> it looks like I had a small snap through there I must have clicked the wrong thing

And what happens if this accident just overwritten all your progress or wiped out all of your personal data without any confirmation at all? How does it feel then?

Cheat Engine process list no need confirmation at all because this is not the end of the world if you accidentally selected the wrong process in the process list because in this case the user can simply spawn again the process list and select the right process that he or she wants to hack but if the user accidentally loads a state instead of saving his progress, especially when both the "Save global state" and "Load global state" are next to each other in the File menu and small in the screen, this is even worse because he or she may lose a lot of progress in this case and he or she will have to remember and repeat all of his or her actions to restore the state and bring back all the progress that he or she lost and waste a lot of precious time by doing all this.

No one will see people doing the mistake that he did in his video because this is very easy to just spawn the process list and select the process that you want to hack (rpcs3 in this case) but even this is a very easy task he still did the mistake because he was always thinking in his mind about what to do next in his tutorial video after doing this very easy task, that is the "Add manual address" step and how to demonstrate and explain this step in his tutorial video, while recording himself (both his screen and his voice) until he didn't notice that he accidentally highlighted Cheat Engine instead of rpcs3 in the process list and then clicked on the Open button because his brain didn't want to pay attention to this easy task at all but concentrating on a harder task like how to demonstrate and explain the next steps in his tutorial video so his tutorial video is clear enough to his watchers.

Inexperienced PCSX-Redux users can also think in their mind what they will do next in the PS1 game that they are playing after doing "Save global state" and as a result they may not notice that they accidentally moved the mouse cursor to point at "Load global state" before the mouse click because doing all what you described is also very easy and no one will bother to pay attention to this.

This is exactly what I am trying to tell you and I can see many people doing this mistake and this is why confirmation is so important where without it accident can cause you to lose a lot of work that you made and even if this won't happen to many people, this still can happen to few of them and why do you want that this will happen to these few people at all?

This already happened to me several times and for him too it seems so!
This happened to me even on PCSX2 1.6.0 because PCSX2 1.6.0 never asks me if I am sure that I want to overwrite an existing p2s file when going to save state to a specified file.

If you had need a proof that I am right and you are wrong then there it is.

Either add this confirmation before "Save global state" and "Load global state" eventually or at least completely remove them from the File menu because leaving them in the File menu in this state can cause many inexperienced users to lose a lot of progress if they will mess up with this badly designed feature. They are unnecessary anyway since the user can save his or her progress to a slot or by using a quick save and load state by pressing F1 and F2 on his or her keyboard or just saving to a memory card like how it is done on a real PlayStation console.

About quick load state I prefer by default to bind it to F3 and not F2, exactly like how it is already done on PCSX2, because both F1 and F2 are next to each other on the keyboard and this can also happen that the user will accidentally press F2 instead of F1 when he or she is going to save his or her progress to a quick state file and think about what he or she will do next in the PS1 game that he or she is playing with PCSX-Redux before the accident. F3 is further away from F1 than F2 on the keyboard decreasing both the chance and probability for a mistake. Personally I will prefer F4 for quick load state, unlike what PCSX2 does, to completely eliminate both the chance and probability for a mistake because F4 is even further away from F1 than F3 on the keyboard.

I will start another discussion about this problem of quick load state in PCSX-Redux.

Nobody will tolerate and suffer from any badly designed GUI applications where mistake can happen and cost him or her a lot of time or money when done.

This is just my opinion, proposal and suggestion based on my bad experience in my life.

This is all what I can tell you about this subject or topic. Now you can do whatever you want about this.

By the way I have already started the discussion concerning the problem of quick load state of PCSX-Redux [here](https://github.com/grumpycoders/pcsx-redux/discussions/1147).

**Reply by ghost** [reply]
> But like, you need to spawn the top menu bar, press the file button, and then click one of those options to do it. I can't see many people doing it accidentally.

I played Final Fantasy XIV/14 yesterday and there is a duty that must be done in order to proceed with the main story.

I tried to register to that duty but accidentally registered to other duty and commenced which I have already done before and I didn't had to repeat that duty at all.

Since players cannot leave any duty without any penalties after commence I had to do it again from the beginning to the end and wasted both time and energy.

This accident happened because the story of Final Fantasy XIV/14 is very interesting and my mind was busy on thinking about what will happen next in that very interesting story before I proceed and didn't notice that I accidentally selected the wrong duty in the duty finder window.

Many PS1 games are not less interesting than Final Fantasy XIV/14, especially Final Fantasy VII/7, VIII/8 and IX/9, and the accident to click "Load global state" instead of "Save global state" in the File menu of PCSX-Redux PlayStation one emulator is much more likely than what you both think and believe.

Also think how much bad it is if this accident happens to someone while he or she is playing a very hard survival horror game with a very interesting story, like Resident Evil, which also greatly limits the player on saving his or her progress to the memory cards and the user will probably want to heavily use the save state feature of the emulator to save his or her progress whenever he or she wants to, if he or she is a casual player, and think about what will happen next in the game and story at the same time and not notice whether or not he or she clicks "Save global state" or "Load global state" at all.

Either ask before "Save global state" and "Load global state" or at least completely remove both "Save global state" and "Load global state" from the File menu of PCSX-Redux PlayStation one emulator because at their current state they can devastate their users and they are unnecessary since their progress can always be saved to a slot, quick and sometimes to the memory cards too which are more safe to use than the global state.

**Reply by NDR008** [reply]
Use a different emulator?

**Reply by ghost** [reply]
> Use a different emulator?

Unfortunately only PCSX-Redux provides all the tools, utilities and features that a cheater and hacker like me needs like Memory Observer Delta-Over-Time search, Memory Read and Write Breakpoints, Memory Editors and etc.

If DuckStation would at least allow me to add and use memory read and write breakpoints like PCSX-Redux does, because DuckStation already has a memory scanner with a feature to freeze any memory address exactly like Cheat Engine, then I would consider to use DuckStation instead.

Even if either Nymashock or Octoshock PS1 cores of Bizhawk had a working debugger then I would use Bizhawk instead of PCSX-Redux but unfortunately neither Nymashock nor Octoshock have a working debugger because Bizhawk also has a memory scanner with the freeze feature exactly like DuckStation does.

Anyway I am just telling you my recent experience and opinions in my life. That's all and if I will have more examples and instances in the future about this subject or topic then I will not hesitate to report them here too to both bold and emphasis the importance of this feature request.

All the collaborators and developers of the PCSX-Redux project are still free to work on what they think is most important now.

Meanwhile I am saving my progress to a slot, neither global (because as I said earlier both save and load are next to each other in the File menu and they both don't ask on click) nor quick (because they are bound to F1 and F2 respectively which are next to each other on the keyboard) nor memory cards (because I can't always save my progress with them), which is more safe while benefiting from all the debug tools and utilities that PCSX-Redux currently provide me and this is okay for me for now.

I am a software quality assurance user and I will keep reporting any new issue here that I will find in the future with PCSX-Redux in order to contribute and improve this excellent PlayStation one emulator.

All what I want to know if you really know the problem that I am describing here?

**Reply by NDR008** [reply]
Don't worry we understood you the first time.
You sometimes clicked on the wrong item from the menu, and it caused you a lot of personal suffering.
But we would prefer you do not keep adding more comments about the same topic.

**Reply by ghost** [reply]
> Don't worry we understood you the first time. You sometimes clicked on the wrong item from the menu, and it caused you a lot of personal suffering. But we would prefer you do not keep adding more comments about the same topic.

I am glad to hear that and I will listen to you and do as you say.
